tools/browser-matcher/matcher-rule.mjs

```diff
@@ -0,0 +1,104 @@
+import { parseSpecString, SPEC_OPERATOR } from './spec-parser.mjs'
+
+export const MATCHER_TYPE = {
+  INCLUDE: 'include',
+  EXCLUDE: 'exclude'
+}
+
+/**
+ * Represents a browser matching rule of the defined type
+ * allowing comparison between two browser specs strings.
+ */
+export default class MatcherRule {
+  #matcherType
+  #specString
+  #browserName
+  #specOperator
+  #browserVersion
+
+  /**
+   * Create a new match rule
+   * @example
+   * // return true if the secondary spec is any version of ie
+   * new MatcherRule(MATCHER_TYPE.INCLUDE, 'ie@*')
+   * @example
+   * // return true if the secondary spec is not any version of ie
+   * new MatcherRule(MATCHER_TYPE.EXCLUDE, 'ie@*')
+   * @example
+   * // return true if the secondary spec is any version of ie > 9
+   * new MatcherRule(MATCHER_TYPE.INCLUDE, 'ie>9')
+   * @param {MATCHER_TYPE} matcherType Type of matching to do between the two browser specs
+   * @param {string} spec The base browser spec string to compare against
+   */
+  constructor (matcherType, specString) {
+    if (!matcherType || !specString) throw new Error('MatcherRule requires a matcherType and specString')
+
+    this.#matcherType = matcherType
+    this.#specString = specString
+
+    const { browserName, specOperator, browserVersion } = parseSpecString(specString)
+
+    this.#browserName = browserName
+    this.#specOperator = specOperator
+    this.#browserVersion = browserVersion
+  }
+
+  /**
+   * Clones the current matcher rule
+   * @returns New instance of the matcher rule with the same criteria
+   */
+  clone () {
+    return new MatcherRule(this.#matcherType, this.#specString)
+  }
+
+  /**
+   * Compares the provided browser spec against the base browser spec used
+   * to construct this matcher rule.
+   * @param {string} spec The browser spec to compare against the base browser spec
+   */
+  test (spec) {
+    const { browserName, _, browserVersion } = parseSpecString(spec)
+
+    if (this.#matcherType === MATCHER_TYPE.EXCLUDE && this.#browserName !== browserName) {
+      return true
+    } else if (this.#matcherType === MATCHER_TYPE.INCLUDE && this.#browserName !== browserName) {
+      return false
+    }
```

Comment:
There's a conflict of default behavior here. For example,

- If I `SpecMatcher().exclude('safari@10')` then `.test('chrome')` it will pass.
- If I `SpecMatcher().exclude('safari@10').include('safari@11')`, then `.test('chrome')` it will **fail**

You should pick a default when this rule doesn't concern `browserName`, regardless if it's an inclusion or exclusion: either return `true` and be inclusive, or return `false` and be exclusive.

Reply:
Those tests make sense. If the matcher is only excluding `safari@10` then `chrome` should pass.  The second test is excluding `safari@10` and only including `safar@11` so unless you pass Safari 11, it shouldn't pass. The default is already set....it matches the rules supplied, period. If you look at a lot of the predefined rules for JIL, they start with `include('*')` or `exclude('*')` for this very reason.